app-android/app/src/main/java/de/tutao/tutanota/Contact.kt

```diff
@@ -37,10 +55,718 @@ class Contact(private val activity: MainActivity) {
+						RawContacts.SOURCE_ID
+				), null, null, null
+		)
+	}
+
+	private fun checkDeletedContact(storedContact: StoredContact, ops: ArrayList<ContentProviderOperation>) {
+		if (storedContact.isDeleted) {
+			val updateDeletedStatusOp = ContentProviderOperation.newUpdate(RawContacts.CONTENT_URI)
+					.withSelection("${ContactsContract.RawContacts._ID} = ?", arrayOf(storedContact.rawId.toString()))
+					.withValue(RawContacts.DELETED, 0)
+					.build()
+			ops += updateDeletedStatusOp
+		} else {
+			Log.d(TAG, "Contact isn't deleted, continuing...")
+		}
+	}
+
+	private fun checkContactDetails(storedContact: StoredContact, serverContact: StructuredContact, ops: ArrayList<ContentProviderOperation>) {
+		if (storedContact.birthday != serverContact.birthday) {
+			checkContactBirthday(storedContact, ops, serverContact)
+		}
+
+		if (storedContact.company != serverContact.company) {
+			checkContactCompany(storedContact, ops, serverContact)
+		}
+
+		if (storedContact.givenName != serverContact.firstName) {
+			val updateNameOp = ContentProviderOperation.newUpdate(ContactsContract.Data.CONTENT_URI)
+					.withSelection(
+							"${ContactsContract.Data.MIMETYPE} = \"${ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE}\" AND ${ContactsContract.Data.RAW_CONTACT_ID} = ?",
+							arrayOf(storedContact.rawId.toString())
+					)
+					.withValue(ContactsContract.CommonDataKinds.StructuredName.GIVEN_NAME, serverContact.firstName)
+					.build()
+			ops += updateNameOp
+		}
+
+		if (storedContact.lastName != serverContact.lastName) {
+			val updateNameOp = ContentProviderOperation.newUpdate(ContactsContract.Data.CONTENT_URI)
+					.withSelection(
+							"${ContactsContract.Data.MIMETYPE} = \"${ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE}\" AND ${ContactsContract.Data.RAW_CONTACT_ID} = ?",
+							arrayOf(storedContact.rawId.toString())
+					)
+					.withValue(ContactsContract.CommonDataKinds.StructuredName.FAMILY_NAME, serverContact.lastName)
+					.build()
+			ops += updateNameOp
+		}
+
+		if (storedContact.nickname != serverContact.nickname) {
+			val updateNicknameOp = ContentProviderOperation.newUpdate(ContactsContract.Data.CONTENT_URI)
+					.withSelection(
+							"${ContactsContract.Data.MIMETYPE} = \"${ContactsContract.CommonDataKinds.Nickname.CONTENT_ITEM_TYPE}\" AND ${ContactsContract.Data.RAW_CONTACT_ID} = ?",
+							arrayOf(storedContact.rawId.toString())
+					)
+					.withValue(ContactsContract.CommonDataKinds.Nickname.NAME, serverContact.nickname)
+					.build()
+			ops += updateNicknameOp
+		}
+	}
+
+	private fun checkContactBirthday(storedContact: StoredContact, ops: ArrayList<ContentProviderOperation>, serverContact: StructuredContact) {
+		// If the birthday wasn't added during contact creation, it's
+		// necessary to add and not just update it
+		if (storedContact.birthday == null) {
+			ops.add(
+					ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
```

Comment:
It would be good to create a facade for `ContentProviderOperation`. For example instead of copy & pasting the operation from `createContact()` to add a birthday, we could call a function called `createBirthdayOperation()` which returns the `ContentProviderOperation` that can be expanded on with `withValue()` as needed.

Reply:
I just noticed `insertAddressOperation()` on the bottom of the file. That is the kind of thing I am referring to.